include/mbedtls/ccm.h

```diff
@@ -284,10 +289,177 @@ int mbedtls_ccm_auth_decrypt( mbedtls_ccm_context *ctx, size_t length,
  */
 int mbedtls_ccm_star_auth_decrypt( mbedtls_ccm_context *ctx, size_t length,
                       const unsigned char *iv, size_t iv_len,
-                      const unsigned char *add, size_t add_len,
+                      const unsigned char *ad, size_t ad_len,
                       const unsigned char *input, unsigned char *output,
                       const unsigned char *tag, size_t tag_len );
 
+/**
+ * \brief           This function starts a CCM encryption or decryption
+ *                  operation.
+ *
+ * \note            This function is not implemented in Mbed TLS yet.
+ *
+ * \param ctx       The CCM context. This must be initialized.
+ * \param mode      The operation to perform: #MBEDTLS_CCM_ENCRYPT or
+ *                  #MBEDTLS_CCM_DECRYPT or #MBEDTLS_CCM_STAR_ENCRYPT or
+ *                  #MBEDTLS_CCM_STAR_DECRYPT.
+ * \param iv        The initialization vector. This must be a readable buffer
+ *                  of at least \p iv_len Bytes.
+ * \param iv_len    The length of the nonce in Bytes: 7, 8, 9, 10, 11, 12,
+ *                  or 13. The length L of the message length field is
+ *                  15 - \p iv_len.
+ * \param total_ad_len   The total length of additional data in bytes.
+ *                       This must be less than `2^16 - 2^8`.
+ * \param plaintext_len  The length in bytes of the plaintext to encrypt or
+ *                       result of the decryption (thus not encompassing the
+ *                       additional data that are not encrypted).
+ *
+ * \return          \c 0 on success.
+ * \return          \#MBEDTLS_ERR_CCM_BAD_INPUT on failure:
+ *                  the context has not been initialized,
+ *                  \p mode is invalid,
+ *                  \p iv_len is invalid (lower than \c 7 or greater than
+ *                  \c 13),
+ *                  \p total_ad_len is greater than \c 0xFF00.
+ */
+int mbedtls_ccm_starts( mbedtls_ccm_context *ctx,
```

Comment:
When looking at CCM in isolation, I liked having `ccm_starts` take both the nonce and the lengths, rather than separate. But [Manuel noticed](https://github.com/ARMmbed/mbedtls/issues/4355#issuecomment-849512636) that this is a poor match for the PSA API which receives the nonce and the lengths in separate functions (for good reasons) and specifies that they can be set in either order (which may have been a bad choice, but it's too late to change now).

So how about removing the lengths from `ccm_starts`, which will align it with `gcm_starts` (but not with chachapoly, which doesn't take a nonce length argument :frowning_face:)? And have a separate function to set the lengths, which can be called after `starts` (it isn't required to support calling it beforehand, but I don't think it would cost anything).

Reply:
Yes sure, good that Manuel had an overall look to the AEAD implementation. Otherwise why not aligning the signature of not mbedtls_chachapoly_starts() with that of the CCM and GCM one as well?  Even if we don't do anything with the new nonce length parameter in mbedtls_chachapoly_starts() as we probably don't have the time for that.

Reply:
I think that would be a good idea, but it's not strictly necessary. 3.0 optional material.

Reply:
I've added mbedtls_ccm_set_lengths() to the API.